vendor/bat-native-ads/src/bat/ads/internal/country_subdivision.cc

```diff
@@ -0,0 +1,195 @@
+  BuildUrl();
+}
+
+CountrySubdivision::~CountrySubdivision() = default;
+
+void CountrySubdivision::FetchCountrySubdivisionIfAllowedForRegion() {
+  auto locale = brave_l10n::LocaleHelper::GetInstance()->GetLocale();
+  const std::string region_code = brave_l10n::GetRegionCode(locale);
+  if (region_code == kCountrySubdivisionAllowedForRegion) {
+    ads_client_->SetSubdivisionAdTargetingRegion(true);
+
+    if (!ads_client_->DidOverrideAdsSubdivision()) {
+      Fetch();
+    }
+  } else {
+    ads_client_->SetSubdivisionAdTargetingRegion(false);
+  }
+}
+
+std::string CountrySubdivision::GetCountrySubdivision() const {
+  return ads_client_->GetCountrySubdivision();
+}
+
+bool CountrySubdivision::IsValidCode(
+    const std::string& country_subdivision_code) const {
+  // TODO(Moritz Haller): Better compare against whitelist?
+  // locale 'US' vs. subdivision 'US-CA'
+  if (country_subdivision_code.length() > 2) {
+    return true;
+  }
+
+  return false;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+void CountrySubdivision::Fetch() {
+  if (retry_timer_.IsRunning()) {
+    return;
+  }
+
+  BLOG(1, "Fetch country subdivision");
+  BLOG(2, "GET " << GETSTATE_PATH);
+
+  auto callback = std::bind(&CountrySubdivision::OnReceived,
+      this, url_, _1, _2, _3);
+  ads_client_->URLRequest(url_, {}, "", "", URLRequestMethod::GET, callback);
+
+  return;
+}
+
+void CountrySubdivision::OnReceived(
+    const std::string& url,
+    const int response_status_code,
+    const std::string& response,
+    const std::map<std::string, std::string>& headers) {
+  BLOG(7, UrlResponseToString(url, response_status_code, response, headers));
+
+  auto should_retry = false;
+
+  if (response_status_code / 100 == 2) {
+    if (!response.empty()) {
+      BLOG(1, "Successfully fetched country subdivision");
+    }
+
+    if (!Process(response)) {
+      BLOG(1, "Failed to parse country subdivision");
+      should_retry = true;
+    }
+  } else if (response_status_code == 304) {
+    BLOG(1, "Country subdivision is up to date");
+  } else {
+    BLOG(1, "Failed to fetch country subdivision");
+
+    should_retry = true;
+  }
+
+  if (should_retry) {
+    Retry();
+    return;
+  }
+
+  retry_timer_.Stop();
+
+  FetchAfterDelay();
+}
+
+bool CountrySubdivision::Process(
+    const std::string& json) {
+  base::Optional<base::Value> root = base::JSONReader::Read(json);
+  if (!root) {
+    return false;
+  }
+
+  base::Value* country_value = root->FindKey("country");
+  if (!country_value) {
+    return false;
+  }
+
+  std::string country;
+  bool parse_country_success = country_value->GetAsString(&country);
+  if (!parse_country_success) {
+    return false;
+  }
+
+  base::Value* region_value = root->FindKey("region");
+  if (!region_value) {
+    return false;
+  }
+
+  std::string region;
+  bool parse_region_success = region_value->GetAsString(&region);
+  if (!parse_region_success) {
+    return false;
+  }
```

Comment:
Going forward we should use `FindStringKey` etc. however we can resolve this as part of https://github.com/brave/brave-browser/issues/8936

Reply:
changed https://github.com/brave/brave-core/pull/5602/files#diff-c33be94bc446f2eda08f8ce0383f81b2R127